code/networking/src/validate_ip/ValidateIp.java

```diff
@@ -0,0 +1,35 @@
+import java.util.Arrays;
+import java.util.regex.Pattern;
+
+public class ValidateIp {
+    public static void main(String[] args) {
+        assert (isValid("172.16.254.1"));
+        assert (!isValid("172.16.254.01"));
+        assert (isValid("2001:0db8:85a3:0000:0000:8a2e:0370:7334"));
+        assert (isValid("2001:db8:85a3:0:0:8A2E:0370:7334"));
+        assert (!isValid("02001:0db8:85a3:0000:0000:8a2e:0370:7334"));
+    }
+
+    private static boolean isValid(String addr) {
+        if (addr.contains(".")) {
+            return isValidv4(addr);
+        }
+        if (addr.contains(":")) {
+            return isValidv6(addr);
+        }
+        return false;
+    }
+
+    private static boolean isValidv4(String addr) {
+        Pattern pattern = Pattern.compile("(?:(?:(?:[1-9][0-9]{0,2})|0)\\.){3}(?:(?:[1-9][0-9]{0,2})|0)");
+        if (!pattern.matcher(addr).matches()) {
+            return false;
+        }
+        return Arrays.stream(addr.split("\\.")).noneMatch(s -> Integer.parseInt(s) > 255);
+    }
+
+    private static boolean isValidv6(String addr) {
+        Pattern pattern = Pattern.compile("(?:(?:(?:[0-9a-fA-F]{1,4})):){7}(?:(?:[0-9a-fA-F]{1,4}))");
```

Comment:
this pattern looks fine with the examples you get, but as mentioned above isn't handling the case with trimmed zeroes, in ipv6 case it'll be two colons ::, i found this [question](https://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses) on stackoverflow is really useful.

Reply:
From the README.md,
> However, we don't replace a consecutive group of zero value with a single empty group using two consecutive colons (::) to pursue simplicity. For example, 2001:0db8:85a3::8A2E:0370:7334 is an invalid IPv6 address.
I think it's being handled.

Reply:
maybe, i need to check it again, but please don't forget about the file name and check the patterns if they working fine.

Reply:
The patterns are working for the couple of examples in the main().
Java code needs the class name to be the same as the file name, and I've never once seen a class name in snakecase.
Still, I can change it for the sake of consistency.